File: frontend/src/types/assetRequest.ts
// 申請画面から POST /api/requests へ送る入力データ
export type AssetRequestCreate = {
  asset_id: number;
  requester_name: string;
  start_date: string;
  end_date: string;
  reason: string;
  quantity: number;
};

// 申請作成・返却・キャンセル API から返る貸出申請データ
export type AssetRequest = AssetRequestCreate & {
  id: number;
  user_id: number;
  status: "pending" | "loaned" | "returned" | "cancelled";
  returned_at: string | null;
  created_at: string;
  updated_at: string;
};

// マイ貸出状況画面で備品名・カテゴリ付きで表示する申請データ
export type ActiveAssetRequest = AssetRequest & {
  asset_name: string;
  asset_category: string;
};

File: backend/app/schemas/asset_request.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.asset_request import AssetRequestStatus


class AssetRequestCreate(BaseModel):
    asset_id: int = Field(ge=1)
    requester_name: str = Field(min_length=1, max_length=120)
    start_date: date
    end_date: date
    reason: str = Field(min_length=1)
    quantity: int = Field(ge=1)

    @model_validator(mode="after")
    def validate_date_range(self) -> "AssetRequestCreate":
        if self.end_date < self.start_date:
            raise ValueError("終了日は開始日以降の日付を指定してください。")
        return self


class AssetRequestRead(BaseModel):
    id: int
    asset_id: int
    user_id: int
    requester_name: str
    start_date: date
    end_date: date
    reason: str
    quantity: int
    status: AssetRequestStatus
    returned_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ActiveAssetRequestRead(AssetRequestRead):
    asset_name: str
    asset_category: str
